resource/graph_node_test.go

```diff
@@ -213,13 +213,41 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
 type someResource struct {
 	resource.Resource
 	closeCap  []interface{}
-	shoudlErr bool
+	shouldErr bool
 }
 
 func (s *someResource) Close(ctx context.Context) error {
 	s.closeCap = append(s.closeCap, ctx.Value("foo"))
-	if s.shoudlErr {
+	if s.shouldErr {
 		return errors.New("bad close")
 	}
 	return nil
 }
+
+type anotherResource struct {
+	resource.Resource
+	CloseFunc func(ctx context.Context) error
+}
+
+// Close calls the injected Close or the real version.
+func (a *anotherResource) Close(ctx context.Context) error {
+	if a.CloseFunc == nil {
+		return errors.New("oops")
+	}
+	return a.CloseFunc(ctx)
+}
+
+func TestClose(t *testing.T) {
+	// Tests that Close does not deadlock by calling graphNode.Close() inside a resource Close().
```

Comment:
[q] How does this mimic the case described in the PR description?

Reply:
this test is a bit narrower - this only tests that graphNode.Close() does not hold the lock while Closing the resource. In the PR description, the issue was that graphNode.Close() held the lock while Closing the resource, and the Close ended up depending on something else that wanted the graphNode lock

Reply:
Cool that makes sense; thanks!